rfcs/20240311-gather-scatter-batching-dims.md

```diff
@@ -0,0 +1,456 @@
+However, unregistered attributes can be discarded at any time and are harder to
+maintain (e.g. if the iota is replaced by another op, who is responsible for
+updating the unregistered attributes).
+
+Another option is to start with `stablehlo.dynamic_slice` and
+`stablehlo.dynamic_update_slice`, which are simpler ops that would suffice for a
+lot of the use cases (but not all) that we've encountered. Currently
+`stablehlo.gather` and `stable.scatter` are used to get a vectorized version of
+`stablehlo.dynamic_slice` and `stablehlo.dynamic_update_slice` respectively.
+However, for the same reason that they can be expressed by gather/scatter, we
+propose to go with the more general solution, that would address all use cases.
+
+## Naming
+
+As indicated above, the proposed naming of the new attributes is inspired by
+`lhs_batching_dims` and `rhs_batching_dims` of `stablehlo.dot_general`. Note
+that the gather op specification already uses the term `batch_dims`
+(`update_scatter_dims` for scatter op), to refer to all dimensions in the result
+tensor that aren't offset dimensions, which have corresponding dimensions in the
+start-indices tensor. The difference is that the proposed dimensions are
+explicit batching dimensions that exist in all operands and result, whereas the
+existing `batch_dims` are implicit batch dimensions (as they are derived by the
+offset dimensions) that exist only in the start-indices and result tensors.
+
+## Proposed Specification
+
+### gather
+
+#### Semantics
+
+Gathers slices from `operand` tensor from offsets specified in `start_indices`
+and produces a `result` tensor.
+
+The following diagram shows how elements in `result` map on elements in
+`operand` using a concrete example. The diagram picks a few example `result`
+indices and explains in detail which `operand` indices they correspond to.
+
+![](images/20240311-gather-scatter-batching-dims/gather.svg)
+
+More formally, `result[result_index] = operand[operand_index]` where:
+
+<!-- markdownlint-disable line-length -->
+* `batch_dims = [d for d in axes(result) and d not in offset_dims]`.
+* `batch_index = result_index[batch_dims...]`.
+* `start_index` is defined as:
+  * `start_indices[bi0, ..., :, ..., biN]` where `bi` are individual elements in
+    `batch_index` and `:` is inserted at the `index_vector_dim` index, if
+    `index_vector_dim` < `rank(start_indices)`.
+  * `[start_indices[batch_index]]` otherwise.
+* For `d_operand` in `axes(operand)`,
+  * `full_start_index[d_operand] = clamp(start_index[d_start], 0, dim(operand, d_operand) - slice_sizes[d_operand])`
+    if `d_operand = start_index_map[d_start]`.
+  * `full_start_index[d_operand] = 0` otherwise.
+* For `d_operand` in `axes(operand)`,
+  * `full_batching_index[d_operand] = batch_index[d_start - (d_start < index_vector_dim ? 0 : 1)]`.
+    if `d_operand = operand_batching_dims[i_batching]`, and `d_start = start_indices_batching_dims[i_batching]`.
```

Comment:
replace `. if` with `, if`
Also can replace the `, and` -> `and` (there are only two clauses)